src/go/k8s/internal/client/factory_test.go

```diff
@@ -0,0 +1,242 @@
+			if tt.Auth != nil {
+				ensureMapAndSetValue(values, "auth", "sasl", map[string]any{
+					"enabled":   true,
+					"secretRef": "users",
+					"users": []map[string]any{{
+						"name":      tt.Auth.Name,
+						"password":  tt.Auth.Password,
+						"mechanism": tt.Auth.Mechanism,
+					}},
+				})
+			}
+
+			// to reduce the bootup time of the cluster
+			ensureMapAndSetValue(values, "statefulset", "replicas", 1)
+			ensureMapAndSetValue(values, "console", "enabled", false)
+			// to keep nodeport services from conflicting
+			ensureMapAndSetValue(values, "external", "enabled", false)
+			ensureMapAndSetValue(values, "image", "tag", "v24.2.2")
+
+			name := fmt.Sprintf("k3s-%d-%d", time.Now().Unix(), suffix.Add(1))
+
+			_, err := helmClient.Install(ctx, "redpandadata/redpanda", helm.InstallOptions{
+				Version:         chartVersion,
+				CreateNamespace: true,
+				Name:            name,
+				Namespace:       name,
+				Values:          values,
+			})
+			require.NoError(t, err)
+
+			t.Run("Cluster", func(t *testing.T) {
+				var cluster redpandav1alpha2.Redpanda
+				cluster.Name = name
+				cluster.Namespace = name
+				cluster.Spec.ClusterSpec = &redpandav1alpha2.RedpandaClusterSpec{}
+
+				data, err := json.Marshal(values)
+				require.NoError(t, err)
+				require.NoError(t, json.Unmarshal(data, cluster.Spec.ClusterSpec))
+
+				kafkaClient, err := factory.KafkaClient(ctx, &cluster)
+				require.NoError(t, err)
+				metadata, err := kadm.NewClient(kafkaClient).BrokerMetadata(ctx)
+				require.NoError(t, err)
+				require.Len(t, metadata.Brokers.NodeIDs(), 1)
+			})
+
+			t.Run("KafkaAPISpec", func(t *testing.T) {
+				var spec redpandav1alpha2.KafkaAPISpec
+				spec.Brokers = []string{fmt.Sprintf("%s-0.%s.%s.svc.cluster.local:9093", name, name, name)}
+				if tt.Auth != nil {
+					require.NoError(t, kubeClient.Create(ctx, &corev1.Secret{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "secret",
+							Namespace: name,
+						},
+						StringData: map[string]string{
+							"password": tt.Auth.Password,
+						},
+					}))
+
+					spec.SASL = &redpandav1alpha2.KafkaSASL{
+						Username: tt.Auth.Name,
+						Password: redpandav1alpha2.SecretKeyRef{
+							Name: "secret",
+							Key:  "password",
+						},
+						Mechanism: redpandav1alpha2.SASLMechanism(tt.Auth.Mechanism),
+					}
+				}
+				if tt.TLS {
+					spec.TLS = &redpandav1alpha2.CommonTLS{
+						CaCert: &redpandav1alpha2.SecretKeyRef{
+							Name: fmt.Sprintf("%s-default-root-certificate", name),
+							Key:  corev1.TLSCertKey,
+						},
+						// This is necessary since we don't have a way of setting a ServerName
+						// through our client initiation code and one or the other must be set
+						// according to go's TLS negotiation code. See:
+						//   https://github.com/golang/go/blob/ab4182251ba57b0d1fe9abbfd604861a6ae75463/src/crypto/tls/handshake_client.go#L46-L48
+						InsecureSkipTLSVerify: true,
```

Comment:
Interestingly, writing these tests with some different configurations of Redpanda to make sure we were pulling a properly initialized client for both `KafkaAPISpec` and a `ClusterRef` made me realize that the `KafkaAPISpec` initialization code might have some issues with TLS-based auth.

Specifically, the underlying go crypto routines get mad because the existing initialization code doesn't set a `ServerName` value on the passed `tls.Config` and if `InsecureSkipTLSVerify` on the `KafkaAPISpec` tls configuration is `false` it throws an error when you try and initialize a client saying that if a `tls.Config` is set, either `ServerName` or `InsecureSkipVerify` [**_must_** be set](https://github.com/golang/go/blob/ab4182251ba57b0d1fe9abbfd604861a6ae75463/src/crypto/tls/handshake_client.go#L46-L48).

We don't have that problem when we're reflecting on the `ClusterRef` since we use the [internal service name](https://github.com/redpanda-data/helm-charts/blob/aca43d8bd857e9d41e68269765fa84f68774203e/pkg/redpanda/client.go#L176) as the `ServerName` field in the passed `tls.Config`, but it seems like if we want to fully support TLS-based stuff in `KafkaAPISpec` we probably should add a `ServerName` field, try and parse one out of some subset of the broker URLs passed to it (probably error prone), or both.

As is, what it means is that TLS-based connections using `KafkaAPISpec` currently _**must always**_ set `InsecureSkipTLSVerify: true`.

Based on any discussion we may have about this, we can file an issue and fix this in a follow-up PR.

Reply:
I would vote to add `ServerName` addition to our CRD.

Reply:
Shouldn't ServerName be getting inferred from the broker hostname(s)? I thought that was [a feature of franz-go](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#DialTLSConfig).

Reply:
Ah, so we might want to copy some of [this chunk of behavior](https://github.com/twmb/franz-go/blob/b77dd13e2bfaee7f5181df27b40ee4a4f6a73b09/pkg/kgo/client.go#L436-L455) from `franz-go`. That said, if you notice the implementation, it only does the `ServerName` injection when `dialFn` is `nil`. We've _always_ overridden `dialFn` through [leveraging `kgo.Dialer`](https://github.com/redpanda-data/redpanda-operator/blob/acc04250c2a14225f885b3c49926036718e965a9/src/go/k8s/internal/controller/redpanda/kgo-config-options.go#L127), so that `nil` check never allows us to use the `ServerName` injection.

If we wanted we could do something a bit trickier like:

1. Introduce a `ServerName` override on the CRD.
2. Add in that `ServerName` injection shim to the `*Spec`-based initialization code.
3. Use the `ServerName` field on the CRD to override what would otherwise be inferred based on the dialed host.

Either way, I'd think that'd be a follow-up PR, but what do y'all think about the above idea?

Reply:
One last update, it actually looks like the `tls.Dialer` we are using in the case of a non-custom dialer (currently in our codebase without the custom dialer implementation I added) actually has the same [sort of behavior as `franz-go`](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/crypto/tls/tls.go;l=140-156). So I think we should actually be fine with our current implementation, the reason this was failing is because of the fact that we pass a dialer in to do the whole port-forwarding business, so we don't use `tls.Dialer` and we don't use that default `dialFn` from `franz-go`. I'll just go ahead and add `ServerName` inferring to the `wrapTLSDialer` code.